### ISSUE TITLE
Far Cry Primal invisible / flickering nature

### ISSUE THREAD

**Issue by SamStrongg** [doitsujin/dxvk#1257]
When I start the game on the latest version of DXVK (1.4.5), trees and plants flicker and the sun is too bright. There is no flickering problem during recording apitrace, which is strange. Everything works well in older versions.

DXVK 1.4.4

![Screenshot_20191126_185555](https://user-images.githubusercontent.com/42712045/69663584-d4cba780-1086-11ea-9523-c68f331e5a86.png)

DXVK 1.4.5

![Screenshot_20191126_185836](https://user-images.githubusercontent.com/42712045/69663629-e8770e00-1086-11ea-8757-e00273e16d8d.png)

### Software information
Far Cry Primal, V-sync off. All details are set to minimum. The issue also occurs in higher settings. Resolution: 1680x1050.

### System information
- GPU: GeForce GTX 750 Ti
- Driver: 440.31
- Wine version: lutris-nofshack-4.19
- DXVK version: 1.4.5

### Apitrace file(s)
https://drive.google.com/open?id=1vESVeEb7nx3Ri-mnLpQoL8e-kLbImaGe

Apitrace with DXVK 1.4.5

### Log files
- d3d11.log: 
[FCPrimal_d3d11.log](https://github.com/doitsujin/dxvk/files/3893559/FCPrimal_d3d11.log)

- dxgi.log: 
[FCPrimal_dxgi.log](https://github.com/doitsujin/dxvk/files/3893560/FCPrimal_dxgi.log)

**Comment by doitsujin** [comment]
This is a known problem with the game and I believe is a game bug, since it seems to rely on uninitialized memory having certain contents. I don't plan to waste any more time on this game.

**Comment by doitsujin** [comment]
See #1155, same issue.

**Comment by SamStrongg** [comment]
So at the moment I can play with the older version of DXVK. This is not a problem for me, but I thought you might be interested. Technical game underdeveloped, but it looks nice and people will buy it;)

**Comment by doitsujin** [comment]
Yeah that mostly happens in the benchmark *in addition* to flickering vegetation. They go hand in hand.

For what it's worth, on my system I can't reproduce the issue with latest `master`, but it **is** broken on 1.4.4.

**Comment by SamStrongg** [comment]
Maybe the location matters, or vram. In time, everyone will forget about this game, nothing to worry about.

**Comment by K0bin** [comment]
Apparently Far Cry 4 & 5 have the same issue.

**Comment by Brn9hrd7** [comment]
In Far Cry 4 the issue started with commit [c7718e5952c06667a8e93fa2e71f4037a462802e](https://github.com/doitsujin/dxvk/commit/c7718e5952c06667a8e93fa2e71f4037a462802e). I don't have Far Cry Primal but since the problem occurred in 1.4.5 first it's probably the same commit.

**Comment by doitsujin** [comment]
I highly doubt the commit in question is actually the *cause* of the bug, but rather triggers some undefined behaviour elsewhere. I still do not understand what the problem is with the Dunia Engine games, and therefore cannot fix it.

**Comment by K0bin** [comment]
Don't know about FC4 but I can confirm that this does **not** fix Far Cry Primal.

**Comment by doitsujin** [comment]
@K0bin what is "this"?

**Comment by K0bin** [comment]
The last commit. Sorry, should've made that a bit more clear.

**Comment by Brn9hrd7** [comment]
Doesn't fix FC4 either

**Comment by Brn9hrd7** [comment]
This issue seems to be very weird indeed, when commenting out everything in InitUavCounter
https://github.com/doitsujin/dxvk/blob/c7718e5952c06667a8e93fa2e71f4037a462802e/src/d3d11/d3d11_initializer.cpp#L52-L64
the issue is still there, but when commenting out the line that calls the InitUavCounter function:
https://github.com/doitsujin/dxvk/blob/c7718e5952c06667a8e93fa2e71f4037a462802e/src/d3d11/d3d11_device.cpp#L431
the issue is gone.

**Comment by Joshua-Ashton** [comment]
For sanity, can you try an MSVC build of `master`?
[msvc.zip](https://github.com/doitsujin/dxvk/files/4022894/msvc.zip)

**Comment by Brn9hrd7** [comment]
When trying these DLLs neither the game nor Uplay launch for me and I get the following error messages:

```
002e:err:module:import_dll Loading library dxgi.dll (which is needed by L"C:\\Program Files (x86)\\Ubisoft\\Ubisoft Game Launcher\\libcef.dll") failed (error c000007b).
002e:err:module:import_dll Loading library d3d11.dll (which is needed by L"C:\\Program Files (x86)\\Ubisoft\\Ubisoft Game Launcher\\libcef.dll") failed (error c000007b).
```

Hope I did everything right, I created a new folder in .local/share/lutris/runtime/dxvk (also the x32 and x64 subfolders) and put the files in there(also changed the DXVK version in Lutris etc). I also tried replacing the DLLs from an already existing folder, but that didn't work either.

**Comment by K0bin** [comment]
MSVC builds dont work either this time.

**Comment by Joshua-Ashton** [comment]
I only provided x64 there

**Comment by doitsujin** [comment]
@Joshua-Ashton are those debug builds? Those don't work on wine because they require debug versions of pretty much all Windows DLLs as well.

**Comment by doitsujin** [comment]
@Brn9hrd7 actually one more idea - can you get rid of the exception hadling, i.e. comment out `try` and the whole `catch` block?

If any of that fixes it, we're bleeding state for some reason, but I don't really see how calling conventions used in private code, or even exception handling, would affect the caller unless either there's a compiler bug or the game makes some really weird assumptions.

**Comment by Brn9hrd7** [comment]
Unfortunately neither adding STDMETHODCALLTYPE to the declaration nor commenting out the exception handling changes anything.

**Comment by K0bin** [comment]
FYI commenting out that method call didn't make a difference with MSVC builds with Primal.

**Comment by Joshua-Ashton** [comment]
> @Joshua-Ashton are those debug builds? Those don't work on wine because they require debug versions of pretty much all Windows DLLs as well.

It was a release build.

**Comment by K0bin** [comment]
The official 1.5.2 builds seem to work again. ¯\\\_(ツ)_/¯

**Comment by Brn9hrd7** [comment]
> The official 1.5.2 builds seem to work again. ¯\_(ツ)_/¯

FC4 has still the flickering issue, unfortunately.

**Comment by kakra** [comment]
With Proton 5.6 GE it doesn't flicker for me but everything is tinted red... And there's no sound except during the intro video.

**Comment by K0bin** [comment]
Related issue and no one knows why that happens either. It happens with some DXVK versions and compiler combinations and not with others.

**Comment by sandsmark** [comment]
I've never used the diffing tool in Ghidra before, but FWIW here's the different disassembly between the code built with and without the call. red outline where the code starts to diverge, before that everything is identical down to register usage (in that function, at least):

![image](https://user-images.githubusercontent.com/25765/91207926-157d4d80-e6f9-11ea-934a-153d6ded3311.png)

**Comment by kakra** [comment]
I thought I'll give this game another try: The red tint is gone now, and performance is much better now. It is even that good that I see 60 fps now instead of 25-ish. But now, vegetation is flickering. If I adjust the settings to stay below 60 fps, there's no flickering. As soon as the fps get above 60 fps (well, it gets to 61 fps), vegetation starts flickering wildly.

It appears that it doesn't matter if vsync is on or off, the frame rate is limited to 60 fps for me. Also, vsync settings don't seem to work at all for me: Either games don't sync at all, or if they sync, vsync lags behind because I see tearing either in the first third of the screen, in the middle, or last third of the screen, sometimes the tearing is moving up and down at little during playing. I'm not sure why that is but whatever it is, it is not syncing correctly for me.

Is it possible that rendering of the scene is not complete when the page flip occurs? And that's an issue once we are above the screen refresh rate? I've seen some other games with flickering lighting/shadows, too, once the fps are near or above the screen refresh rate so there might be some syncing missing in the code?

The brightness/red tint may be a similar issue: If I run in very high settings / high resolution and manage to get down to 10-15 fps, I see the red tinting and high brightness for a few seconds during the benchmark. But when the game sits locked at 60 fps, the red tint and brightness goes away during the first few frames rendered.

For FC Primal, I've only run the benchmark which MAY ignore vsync anyways - I'm not sure. I'm currently trying to figure out why it ignores my gamepad... But that's a different issue not belonging here.

Using: Proton Experimental (I think it ships latest DXVK)

**Comment by OrangeCatSeth** [comment]
This is fixed in Proton-6.0-GE-1, at least for far cry 4. I assume primal is also fixed, but don't own the game to test it.

**Comment by K0bin** [comment]
This should hopefully be fixed by #2137
